torchaudio_contrib/layers.py

```diff
@@ -1,142 +1,241 @@
+            stft_out (Tensor): (channel, time, freq, complex)
+                or (batch, channel, time, freq, complex).
         """
 
-        if self.pad > 0:
-            with torch.no_grad():
-                x = F.pad(x, (self.pad, self.pad), "constant")
+        # use registered window so have to use _stft
+        stft_out = _stft(
+            x,
+            self.n_fft,
+            self.hop_length,
+            window=self.window,
+            pad=self.pad,
+            pad_mode=self.pad_mode,
+            **self.kwargs)
+
+        return stft_out
+
+    def __repr__(self):
+        param_str = '(n_fft={}, hop_length={}, len_win={}, pad={})'.format(
+            self.n_fft, self.hop_length, self.window.size(0), self.pad)
+        return self.__class__.__name__ + param_str
+
+
+class ComplexNorm(nn.Module):
+    """
+    Wrap torchaudio_contrib.complex_norm in an nn.Module.
+    """
+
+    def __init__(self, power=1.0):
+        super(ComplexNorm, self).__init__()
+        self.power = power
 
-        spec = spectrogram(x,
-                    n_fft=self.n_fft,
-                    hop=self.hop,
-                    window=self.window,
-                    **self.kwargs)
+    def forward(self, x):
+        return complex_norm(x, self.power)
+
+
+class Filterbank(_ModuleNoStateBuffers):
+
+    def __init__(self):
+        super(Filterbank, self).__init__({'fb': self._build_fb()})
+
+    def _build_fb(self):
+        raise NotImplementedError
+
+    def forward(self, x):
+        """
+        Args:
+            x (Tensor): (channel, time, freq) or (batch, channel, time, freq).
 
-        return spec
+        Returns:
+            ret (Tensor): freq -> fb.size(0)
+        """
+        return torch.matmul(x.transpose(-2, -1), self.fb).transpose(-2, -1)
 
 
-class Melspectrogram(Spectrogram):
+class MelFilterbank(Filterbank):
     """
-    Module that outputs the mel-spectrogram (transform on the spectrogram
-    to better represent human perception) of an audio signal.
+    Convert a spectrogram into a mel frequency spectrogram using
+    a filterbank.
 
     Args:
-     * n_mels: int > 0 
-       -  Number of mel bands.
-       -  Default: 128
-     * sr: int > 0
-       -  Sampling rate of the audio signal. This may not be the same in all samples (?)
-       -  Default: 44100
-    * f_min: float > 0
-       -  Lowest freq. in Hz
-       -  Default: 0.
-    * f_max: float > 0
-       -  Highest freq. in Hz
-       -  Default: None (then use sr / 2.0)
-    * args: 
-       -  Positional arguments for Spectrogram
-    * kwargs: 
-       -  Keyword arguments for Spectrogram
+        n_mels (int): number of mel bins. Defaults to 128.
+        sr (int): sample rate of audio signal. Defaults to 44100.
+        f_max (float, optional): maximum frequency. Defaults to sr // 2.
+        f_min (float): minimum frequency. Defaults to 0.
+        n_stft (int, optional): number of filter banks from stft.
+            Defaults to 2048//2 + 1.
```

Comment:
>I always find shortened parameter names a bit cryptic

Yeah, I think we should open this up to discussion. 

>I also find it convenient sometimes to make the sample rate optional, and expect min_freq and max_freq to be given relative to the Nyquist rate if the sample rate is not given; 

But we do the opposite right, because we have the default value of `f_max` be `sr//2`. So this would be switching up these two?

Reply:
> So this would be switching up these two?

Depends on what users want. To get to the current behavior, we would default `sr=2` and `f_max=1.0`, and users wouldn't need to specify anything at all. That's how `scipy.signal` does it. If they specify a sample rate, they will also need to specify f_max (otherwise there's no reason to specify the sample rate).
 If `sr=2` is too confusing (although it's the correct default to make `f_min` and `f_max` relative to Nyquist), we could also default `sr=None`. And if we want people to be able to specify the sample rate for no reason (or to only change `f_min`) and still have `f_max` default to Nyquist, then we'd also need to default `f_max=None`.